plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java

```diff
@@ -70,17 +73,18 @@ public HdfsRepository(
             throw new IllegalArgumentException("No 'uri' defined for hdfs snapshot/restore");
         }
         uri = URI.create(uriSetting);
-        if ("hdfs".equalsIgnoreCase(uri.getScheme()) == false) {
+        if (uri.getScheme() == null || allowedSchemes.contains(uri.getScheme().toLowerCase(Locale.ROOT)) == false) {
             throw new IllegalArgumentException(
                 String.format(
                     Locale.ROOT,
-                    "Invalid scheme [%s] specified in uri [%s]; only 'hdfs' uri allowed for hdfs snapshot/restore",
+                    "Invalid scheme [%s] specified in uri [%s]; only one of %s uris allowed for hdfs snapshot/restore",
                     uri.getScheme(),
-                    uriSetting
+                    uriSetting,
+                    String.join(", ", allowedSchemes)
                 )
             );
         }
-        if (Strings.hasLength(uri.getPath()) && uri.getPath().equals("/") == false) {
+        if (uri.getScheme().equals("hdfs") && Strings.hasLength(uri.getPath()) && uri.getPath().equals("/") == false) {
```

Comment:
Good point, that's a bit of wishful thinking on my part. You see, I have separate PR for Hadoop: https://github.com/apache/hadoop/pull/5447 that would utilize base path in webhdfs-scheme URIs. I was actually testing this PR combined with my Hadoop PR at one point.


For some context: `webhdfs` is using regular HTTP(S) requests for interacting with filesystem. The requests are being made to base scheme URI + a [hardcoded prefix](https://github.com/trakos/hadoop/blob/162288bc0af944f116a4dd73e27f4676a204d9e9/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java#L150), `/webhdfs/v1`. However, there are some webhdfs proxies, most notably [Apache Knox](https://knox.apache.org/), that have a different API path. It's also easy to set up a custom proxy, since it's just HTTP.

For Apache Knox, the default path is `/gateway/<gateway-name>/webhdfs/v1`. Somewhat ironically, Apache Hadoop's client does not support this custom path currently, unlike most third party clients. In that PR https://github.com/apache/hadoop/pull/5447 I've linked to before I proposed to add support for it by utilizing the path in connection URI. So, while it's not the same as filesystem base path, there would be a use-case where user would want to provide a separate path in connection URI, for example  `webhdfs://hostname:9870/gateway/myname/webhdfs/v1/`. It would be only path to API, different from a `path` option.

Anyway, I can make this validation apply to (s)webhdfs schemes for now if you'd prefer, please let me know. We could then perhaps revisit if Hadoop would accept the change.